In [1]:
import pandas as pd

# 1. Import des données

In [2]:
df_irve = pd.read_csv(
 'https://www.data.gouv.fr/api/1/datasets/r/eb76d20a-8501-400e-b336-d85724de5435'
)

print(f"Shape : {df_irve.shape}")
df_irve.head()

/tmp/ipykernel_89352/3761374698.py:1: DtypeWarning: Columns (12,18,19,20,21,22,24,25,29,33) have mixed types. Specify dtype option on import or set low_memory=False.
  df_irve = pd.read_csv(


Shape : (203199, 52)


,nom_amenageur,siren_amenageur,contact_amenageur,nom_operateur,contact_operateur,telephone_operateur,nom_enseigne,id_station_itinerance,id_station_local,nom_station,...,datagouv_resource_id,datagouv_organization_or_owner,created_at,consolidated_longitude,consolidated_latitude,consolidated_code_postal,consolidated_commune,consolidated_is_lon_lat_correct,consolidated_is_code_insee_verified,consolidated_is_code_insee_modified
0,ChargePoint,NaN,info@chargepoint.com,ChargePoint,info@chargepoint.com,NaN,ACU_Poste_De_Garde_Haguenau,ATHTBE1004017,ATHTBE1004017,ACU_Poste_De_Garde_Haguenau,...,b11113db-875d-41c7-8673-0cf8ad43e917,eco-movement,2023-06-28T11:46:08.539000+00:00,7.762694,48.825613,NaN,NaN,False,False,False
1,ChargePoint,NaN,info@chargepoint.com,ChargePoint,info@chargepoint.com,NaN,ACU_Poste_De_Garde_Haguenau,ATHTBE1004018,ATHTBE1004018,ACU_Poste_De_Garde_Haguenau,...,b11113db-875d-41c7-8673-0cf8ad43e917,eco-movement,2023-06-28T11:46:08.539000+00:00,7.762694,48.825613,NaN,NaN,False,False,False
2,ChargePoint,NaN,info@chargepoint.com,ChargePoint,info@chargepoint.com,NaN,ACU_Poste_De_Garde_Haguenau,ATHTBE1004019,ATHTBE1004019,ACU_Poste_De_Garde_Haguenau,...,b11113db-875d-41c7-8673-0cf8ad43e917,eco-movement,2023-06-28T11:46:08.539000+00:00,7.762694,48.825613,NaN,NaN,False,False,False
3,ChargePoint,NaN,info@chargepoint.com,ChargePoint,info@chargepoint.com,NaN,ACU_Poste_De_Garde_Haguenau,ATHTBE1004020,ATHTBE1004020,ACU_Poste_De_Garde_Haguenau,...,b11113db-875d-41c7-8673-0cf8ad43e917,eco-movement,2023-06-28T11:46:08.539000+00:00,7.762694,48.825613,NaN,NaN,False,False,False
4,ChargePoint,NaN,info@chargepoint.com,ChargePoint,info@chargepoint.com,NaN,HAG_P22_Slave_3,ATHTBE1004957,ATHTBE1004957,HAG_P22_Slave_3,...,b11113db-875d-41c7-8673-0cf8ad43e917,eco-movement,2023-06-28T11:46:08.539000+00:00,7.761841,48.827040,NaN,NaN,False,False,False


Ce premier jeu de données est la base nationale des IRVE (Infrastructures de Recharge pour Véhicules Électriques).
Cette table comporte 210129 lignes pour 52 variables.

In [3]:
df_ve = pd.read_csv(
 'https://www.data.gouv.fr/api/1/datasets/r/90e0d717-deda-4bdc-9987-f82faac5bc93',
 encoding='latin-1'
)

print(f"Shape : {df_ve.shape}")
df_ve.head()

Shape : (703545, 8)


/tmp/ipykernel_89352/3535292153.py:1: DtypeWarning: Columns (0,2) have mixed types. Specify dtype option on import or set low_memory=False.
  df_ve = pd.read_csv(


,CODGEO,LIBGEO,EPCI,LIBEPCI,DATE_ARRETE,NB_VP_RECHARGEABLES_EL,NB_VP_RECHARGEABLES_GAZ,NB_VP
0,01022,ARTEMARE,200040350.0,CC Bugey Sud,2025-09-30,26,0,1446
1,01022,ARTEMARE,200040350.0,CC Bugey Sud,2025-06-30,27,0,1437
2,01022,ARTEMARE,200040350.0,CC Bugey Sud,2025-03-31,20,0,1425
3,01022,ARTEMARE,200040350.0,CC Bugey Sud,2024-12-31,20,0,1416
4,01022,ARTEMARE,200040350.0,CC Bugey Sud,2024-09-30,21,0,1406


Ce second jeu de données contient des informations sur les voitures particulières immatriculées par commune et par type de recharge.
Cette table comporte 703545 lignes pour 8 variables.

In [4]:
df_revenus = pd.read_csv(
 'https://www.data.gouv.fr/api/1/datasets/r/516130bc-4dcb-47f5-8347-ae96553c43ab',sep=';'
)

print(f"Shape : {df_revenus.shape}")
df_revenus.head()

Shape : (34926, 57)


/tmp/ipykernel_89352/2507393620.py:1: DtypeWarning: Columns (1) have mixed types. Specify dtype option on import or set low_memory=False.
  df_revenus = pd.read_csv(


,Nom géographique GMS,Code géographique,Libellé géographique,[DISP] Nbre de ménages fiscaux,[DISP] Nbre de personnes dans les ménages fiscaux,[DISP] Nbre d'unités de consommation dans les ménages fiscaux,[DISP] 1ᵉʳ quartile (€),[DISP] Médiane (€),[DISP] 3ᵉ quartile (€),[DISP] Écart interquartile (€),...,[DEC] 9ᵉ décile (€),[DEC] Rapport interdécile D9/D1,[DEC] S80/S20,[DEC] Iice de Gini,[DEC] Part des revenus d’activité (%),[DEC] dont part des salaires et traitements (%),[DEC] dont part des iemnités de chômage (%),[DEC] dont part des revenus des activités non salariées (%),"[DEC] Part des pensions, retraites et rentes (%)",[DEC] Part des autres revenus (%)
0,L'Abergement-Clémenciat,01001,L'Abergement-Clémenciat,346.0,895.0,590.8,NaN,25820.0,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
1,L'Abergement-de-Varey,01002,L'Abergement-de-Varey,115.0,266.0,181.0,NaN,24480.0,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
2,Ambérieu-en-Bugey,01004,Ambérieu-en-Bugey,6855.0,15092.0,10398.2,15800.0,21660.0,28430.0,12630.0,...,39380.0,"5,6",8.1,0.368,67.9,61.5,3.2,3.2,24.4,7.7
3,Ambérieux-en-Dombes,01005,Ambérieux-en-Dombes,800.0,2028.0,1329.7,20010.0,24610.0,31180.0,11170.0,...,43610.0,"3,2",4.4,0.298,73.2,67.2,2.2,3.8,19.7,7.1
4,Ambléon,01006,Ambléon,51.0,107.0,76.6,NaN,24210.0,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN


Ce dernier jeu de données contient des informations sur les voitures particulières immatriculées par commune et par type de recharge.
Cette table comporte 34926 lignes pour 57 variables.

# 2. Choix de la variable de jointure

In [5]:
list(df_irve.columns)

['nom_amenageur',
 'siren_amenageur',
 'contact_amenageur',
 'nom_operateur',
 'contact_operateur',
 'telephone_operateur',
 'nom_enseigne',
 'id_station_itinerance',
 'id_station_local',
 'nom_station',
 'implantation_station',
 'adresse_station',
 'code_insee_commune',
 'coordonneesXY',
 'nbre_pdc',
 'id_pdc_itinerance',
 'id_pdc_local',
 'puissance_nominale',
 'prise_type_ef',
 'prise_type_2',
 'prise_type_combo_ccs',
 'prise_type_chademo',
 'prise_type_autre',
 'gratuit',
 'paiement_acte',
 'paiement_cb',
 'paiement_autre',
 'tarification',
 'condition_acces',
 'reservation',
 'horaires',
 'accessibilite_pmr',
 'restriction_gabarit',
 'station_deux_roues',
 'raccordement',
 'num_pdl',
 'date_mise_en_service',
 'observations',
 'date_maj',
 'cable_t2_attache',
 'last_modified',
 'datagouv_dataset_id',
 'datagouv_resource_id',
 'datagouv_organization_or_owner',
 'created_at',
 'consolidated_longitude',
 'consolidated_latitude',
 'consolidated_code_postal',
 'consolidated_commune',
 '

In [6]:
list(df_ve.columns)

['CODGEO',
 'LIBGEO',
 'EPCI',
 'LIBEPCI',
 'DATE_ARRETE',
 'NB_VP_RECHARGEABLES_EL',
 'NB_VP_RECHARGEABLES_GAZ',
 'NB_VP']

In [7]:
list(df_revenus.columns)

['Nom géographique GMS',
 'Code géographique',
 'Libellé géographique',
 '[DISP] Nbre de ménages fiscaux',
 '[DISP] Nbre de personnes dans les ménages fiscaux',
 "[DISP] Nbre d'unités de consommation dans les ménages fiscaux",
 '[DISP] 1ᵉʳ quartile (€)',
 '[DISP] Médiane (€)',
 '[DISP] 3ᵉ quartile (€)',
 '[DISP] Écart interquartile (€)',
 '[DISP] 1ᵉʳ décile (€)',
 '[DISP] 2ᵉ décile (€)',
 '[DISP]3ᵉ décile (€)',
 '[DISP] 4ᵉ décile (€)',
 '[DISP] 6ᵉ décile (€)',
 '[DISP] 7ᵉ décile (€)',
 '[DISP] 8ᵉ décile (€)',
 '[DISP] 9ᵉ décile (€)',
 '[DISP] Rapport interdécile 9ᵉ décile/1ᵉʳ decile',
 '[DISP] S80/20',
 '[DISP] Iice de Gini',
 '[DISP] Part des revenus d’activité (%)',
 '[DISP] dont part des salaires et traitements(%)',
 '[DISP] dont part des iemnités de chômage (%)',
 '[DISP] dont part des revenus des activités non salariées (%)',
 '[DISP] Part des pensions, retraites et rentes (%)',
 '[DISP] Part des revenus du patrimoine et autres revenus (%)',
 "[DISP] Part de l'ensemble des prestat

Nous remarquons que ces 3 jeu de données contiennent une variable indiquant le code géographique mais sous 3 noms différents :
- `code_insee_commune` pour df_irve
- `CODGEO` pour df_ve
- `Code géographique` pour df_revenus

# 3. Préparation de la variable de jointure

In [8]:
from src.preparation_data import diagnostic_cle_jointure

In [9]:
diagnostic_cle_jointure(df_irve,"code_insee_commune","df_irve")
diagnostic_cle_jointure(df_ve,"CODGEO","df_ve")
diagnostic_cle_jointure(df_revenus,"Code géographique","df_revenus")

--- Diagnostic clé de jointure : code_insee_commune (df_irve) ---
Type de la colonne           : object
Nb lignes total              : 203199
Nb valeurs manquantes (NaN)  : 57886
Nb valeurs non nulles        : 145313
Nb valeurs uniques (total)   : 11046
Nb valeurs uniques (sans NaN): 11046

Distribution des longueurs :
code_insee_commune
3         4
5    107325
6      1133
7     36851
Name: count, dtype: int64

!!!! Clé NON unique !!!!
!!!! Présence de valeurs manquantes !!!!
--------------------------------------------------

--- Diagnostic clé de jointure : CODGEO (df_ve) ---
Type de la colonne           : object
Nb lignes total              : 703545
Nb valeurs manquantes (NaN)  : 0
Nb valeurs non nulles        : 703545
Nb valeurs uniques (total)   : 35206
Nb valeurs uniques (sans NaN): 35206

Distribution des longueurs :
CODGEO
4     35747
5    667798
Name: count, dtype: int64

!!!! Clé NON unique !!!!
--------------------------------------------------

--- Diagnostic clé de jointur

Ce diagnostic met en évidence plusieurs points d’attention :

- Présence de valeurs manquantes dans `code_insee_commune` (df_irve)  
- Présence de longueurs de codes incohérentes :
  - `code_insee_commune` (df_irve) contient des codes de longueur 3, 6 et 7
  - `CODGEO` (df_ve) contient des codes de longueur 4
- Les variables de jointure ne sont pas uniques :
  - `code_insee_commune` (df_irve)
  - `CODGEO` (df_ve)

Les clés de jointure présentent des problèmes de qualité et de format.  
Il est nécessaire de les nettoyer et les uniformiser (type, format, gestion des valeurs manquantes, unicité) avant d’effectuer la jointure.

#### a. repérons les problèmes de longueurs de codes incohérentes

In [10]:
taile_louche = df_irve[
    df_irve["code_insee_commune"].notna() &
    (df_irve["code_insee_commune"].astype(str).str.len() == 7)
]
taile_louche["code_insee_commune"]

65600     59378.0
65601     59378.0
65602     59378.0
65603     59378.0
65604     59378.0
           ...   
163835    34172.0
163836    34172.0
163837    34172.0
163838    13028.0
163839    13028.0
Name: code_insee_commune, Length: 36851, dtype: object

longueur 3 c'est les 0.0
longueur 6 c'est les codes à 4 chiffres suivis de .0
longueur 7 c'est les codes à 5 chiffres suivis de .0

In [11]:
taile_louche = df_ve[
    df_ve["CODGEO"].notna() &
    (df_ve["CODGEO"].astype(str).str.len() == 4)
]
taile_louche["CODGEO"]

109071    1005
109072    1005
109073    1005
109074    1005
109075    1005
          ... 
555571    9340
555572    9340
555573    9340
555574    9340
555575    9340
Name: CODGEO, Length: 35747, dtype: object

longueur 4 c'est les codes à 4 chiffres

#### b. uniformisons ces codes géographiques pour qu'ils soient tous à 5 chiffres

In [12]:
from src.preparation_data import nettoyer_code_insee

In [13]:
df_irve["code_insee_commune"] = df_irve["code_insee_commune"].apply(nettoyer_code_insee)
df_ve["CODGEO"] = df_ve["CODGEO"].apply(nettoyer_code_insee)
df_revenus["Code géographique"] = df_revenus["Code géographique"].apply(nettoyer_code_insee)

In [14]:
diagnostic_cle_jointure(df_irve,"code_insee_commune","df_irve")
diagnostic_cle_jointure(df_ve,"CODGEO","df_ve")

--- Diagnostic clé de jointure : code_insee_commune (df_irve) ---
Type de la colonne           : object
Nb lignes total              : 203199
Nb valeurs manquantes (NaN)  : 57890
Nb valeurs non nulles        : 145309
Nb valeurs uniques (total)   : 9308
Nb valeurs uniques (sans NaN): 9308

Distribution des longueurs :
code_insee_commune
5    145309
Name: count, dtype: int64

!!!! Clé NON unique !!!!
!!!! Présence de valeurs manquantes !!!!
--------------------------------------------------

--- Diagnostic clé de jointure : CODGEO (df_ve) ---
Type de la colonne           : object
Nb lignes total              : 703545
Nb valeurs manquantes (NaN)  : 0
Nb valeurs non nulles        : 703545
Nb valeurs uniques (total)   : 35197
Nb valeurs uniques (sans NaN): 35197

Distribution des longueurs :
CODGEO
5    703545
Name: count, dtype: int64

!!!! Clé NON unique !!!!
--------------------------------------------------



#### c. diminuons le nombre de valeurs manquantes en utilisant la latitude et longitude

In [15]:
from src.preparation_data import creer_gdf_irve, charger_communes, joindre_communes, ajouter_codes_geo

In [16]:
gdf_irve = creer_gdf_irve(df_irve, "consolidated_longitude", "consolidated_latitude")
gdf_communes = charger_communes()
gdf_result = joindre_communes(gdf_irve, gdf_communes)
df_irve = ajouter_codes_geo(df_irve, gdf_result)

There was an error while reading the file from the URL: https://minio.lab.sspcloud.fr/projet-cartiflette/production/provider=IGN/dataset_family=ADMINEXPRESS/source=EXPRESS-COG-CARTO-TERRITOIRE/year=2022/administrative_level=COMMUNE/crs=4326/DEPARTEMENT=20/vectorfile_format=geojson/territory=metropole/simplification=0/raw.geojson
Error message: '/vsimem/pyogrio_1f7e31a1b06d463db8fc6fd6f91b9db0' not recognized as being in a supported file format.; It might help to specify the correct driver explicitly by prefixing the file path with '<DRIVER>:', e.g. 'CSV:path'.


Nous avons maintenant 3 colonnes concernant le code géographique dans la base de données irve :
- `code_insee_commune` : la variable initiale
- `code_geo_manquant` : la variable initiale complétée grace à la latitude et la longitude pour les valeurs manquantes
- `code_geo_total` : la variable recalculant tous les codes géographiques à partir de la latitude et la longitude, sans prendre en compte la variable initiale `code_insee_commune`

In [17]:
from src.preparation_data import compter_valeurs_manquantes, compter_uniques

In [ ]:
colonnes_irve = ["code_insee_commune", "code_geo_manquant", "code_geo_total"]

# Valeurs manquantes
print("Valeurs manquantes :")
print(compter_valeurs_manquantes(df_irve, colonnes_irve))

# Valeurs uniques
print("\nValeurs uniques :")
print(compter_uniques(df_irve, colonnes_irve))

Valeurs manquantes :
{'code_insee_commune': np.int64(57890), 'code_geo_manquant': np.int64(1240), 'code_geo_total': np.int64(1768)}

Valeurs uniques :
{'code_insee_commune': 9308, 'code_geo_manquant': 11788, 'code_geo_total': 10253}


In [19]:
set_irve_manquant = set(df_irve["code_geo_manquant"].dropna())
set_irve_total = set(df_irve["code_geo_total"].dropna())
set_ve = set(df_ve["CODGEO"].dropna())
set_revenus = set(df_revenus["Code géographique"].dropna())

In [20]:
print("Différences avec VE :")
print(len(set_irve_manquant - set_ve))
print(len(set_irve_total - set_ve))

print("\nDifférences avec revenus :")
print(len(set_irve_manquant - set_revenus))
print(len(set_irve_total - set_revenus))

print("\nDifférences VE vs IRVE :")
print(len(set_ve - set_irve_manquant))
print(len(set_ve - set_irve_total))

print("\nDifférences revenus vs IRVE :")
print(len(set_revenus - set_irve_manquant))
print(len(set_revenus - set_irve_total))

Différences avec VE :
817
0

Différences avec revenus :
828
3

Différences VE vs IRVE :
24226
24944

Différences revenus vs IRVE :
23966
24676


vérifier qu'il n'y a pas les arrondissements dans code_insee_commune ou juste pas à jour ou faux
apporter une correction à ceux qui se trouve à la fois dans len(set(df_irve["code_geo_manquant"]) - set(df_revenus["Code géographique"])) et dans len(set(df_irve["code_geo_manquant"]) - set(df_ve["CODGEO"]))

In [21]:
codes_manquants_communs = (set_irve_manquant - set_revenus) & (set_irve_manquant - set_ve)

print("Nombre de codes manquants dans les deux :")
print(len(codes_manquants_communs))

Nombre de codes manquants dans les deux :
817


garder la variable code_geo_manquant mais pour tout cela prendre la valeur de la variable code_geo_total